### ISSUE TITLE
Exoplayer fails to initialize after multiple release timeouts

### ISSUE THREAD

**Issue by denis-bezrukov** [google/ExoPlayer#9806]
Background: 
We heavily use exoplayer in our app - we have 10+ various screens that play a video inside. We do not use shared instance of exoplayer, so each screen may init its own player for the playback thus we very often release them (as well as underlying MediaCodecs). To catch any missed `player.release` and avoid perf regressions (e.g. playing a video in invisible fragment), we have a global debug check that no more than 2 players are in prepared state, and no more than 1 player is in playing state at the same time.
  
We're on a custom hardware (android 6). This device has mediatek's AVC/HEVC codecs (we have mixed content, but moving forward with hevc)

Recently we started seeing increased amount of such playback erros (we use custom 2s timeout comparing to 500ms in the library) 
<details>

```
com.google.android.exoplayer2.ExoTimeoutException: Player release timed out.
    at com.google.android.exoplayer2.ExoPlayerImpl.lambda$release$5
    at com.google.android.exoplayer2.-$$Lambda$ExoPlayerImpl$eZVQ1P4AuRBXX3IBVzj-JmjEs8k.invoke(_:0)
    at com.google.android.exoplayer2.util.ListenerSet$ListenerHolder.invoke
    at com.google.android.exoplayer2.util.ListenerSet.lambda$queueEvent$0
    at com.google.android.exoplayer2.util.-$$Lambda$ListenerSet$NbKDn9xtItiyMgYZmjIx_Sv1FFQ.run(_:0)
    at com.google.android.exoplayer2.util.ListenerSet.flushEvents(_:191)
    at com.google.android.exoplayer2.util.ListenerSet.sendEvent
    at com.google.android.exoplayer2.ExoPlayerImpl.release(_:854)
    at com.google.android.exoplayer2.SimpleExoPlayer.release(_:1674)
```
</details>
Sometimes it fails with timeout during detaching a surface. Probably it's related to upgrade from 2.13 to 2.15.1 (Probably, because we continue release new features related to videos, so it could be related to increased amount of `prepare` -> `release` cycles), but I didn't find anything relayted to it in the changelog / sourcecode

To reproduce it I built a screen containing 9 PlayerViews (3x3 grid) and repeatedly playback a video in these views. Each 10-120s (randomly) I restart the playback (release all the players, and recreat them again). Sometimes issue start to happen quickly, sometimes it may take a time (like 30 minutes+) to start happening. 
After a few first timeouts it simply can't init a new player (with no error, but sometimes it seems it can thow smth like `MediaCodecVideoRenderer error, index=0, format=Format`), but if attempts to start a playback continues (in user case they navigate to and from the screen with player) device may enter into non-recoverable ANR (touches won't help, ANR dialog doesn't appear, only power-off may help). 
I tried to minimize this reproducer and move it to exo's demo app and end up with https://github.com/denis-bezrukov/ExoPlayer/commit/cc6fb7f65e865e1a994b1f3a62aa0065797e5968 (To ensure `.release()` is called for every player this sample has counter that incremented each time player created, and decremented each time `.release()` is called. Also don't care onStop is not handled and coroutine is not cancelled - steps do not include closing this activity)

I will also submit a anr trace example to the dev.exoplayer@gmail.com

I searched through the issues, but all I found is related to DRM or Dummy Surface workaroun. Our codecs don't work nicely with Dummy surface, but overriding codecNeedsSetOutputSurfaceWorkaround (since it's custom device, it is not blacklisted in the library) to true didn't seem to help in this case and the issue is still reproducible. 

I don't want to put a lot of random info here, so please let me know if I can collect any useful info that may help to track down the issue.

**Comment by marcbaechinger** [comment]
Seems to me like after a couple of failed releases, the number of available codecs the device has is used and no further player can be created because of lack of available codecs.

@christosts May I ask you to look into this. Any ideas how Denis could find an approach that helps?

**Comment by adamvernier** [comment]
We use exo player heavily for a digital sign product that plays videos in loops. We have seen this problem on multiple devices that we work with. It seems to happen more frequently the more the videos are started/stopped. 
In particular we see logs like this in onPlayerError() when trying to start a video:
Video Error errorCodeName=ERROR_CODE_TIMEOUT message=Unexpected runtime error
and sometimes...
Video Error errorCodeName=ERROR_CODE_DECODER_INIT_FAILED message=MediaCodecVideoRenderer error, index=0, format=Format(1, null, null, video/x-vnd.on2.vp9, null, -1, en, [3840, 2160, -1.0], [-1, -1]), format_supported=YES
Renderer diagnostic = android.media.MediaCodec.error_1100

When in this state the device will no longer play videos and will report the 1100 error. I think this is the codec saying it doesn't have enough memory and this is the same error we see when trying to play a very large video that is not supported by the device (larger than 4k for example). 

I've tried to validate and be *sure* that i'm always calling releasePlayer(). So, my thought is that somehow I'm still not calling it in some situations or there is some sort of failure to release it. 

I have found only a device reboot helps when this happens. 

Is there any other way to release players that might have not released? Is there a way to check/verify that the release has occurred? Any further help with this issue would be greatly appreciated. thanks!

**Comment by christosts** [comment]
@denis-bezrukov I will look at your code sample in more detail, but looking quickly on the anr trace example you sent us, I see player is blocked trying to create `DummySurfaces`. You mentioned that your device is not part of the device list hard-coded in the library, and in general your custom device does not work well with the workaround. May I ask if the submitted trace is taken with or without the surface workaround enabled by you?

**Comment by denis-bezrukov** [comment]
@christosts iirc traces were taken without surface workaround

**Comment by christosts** [comment]
@adamvernier can you please file a separate issue, considering adding the information requested in the[ issue template](https://github.com/google/ExoPlayer/blob/release-v2/.github/ISSUE_TEMPLATE/bug.md)? So far, this issue seems to be related to a custom device with Android 6 and running ExoPlayer 2.15.1 whereas your setup is likely to be different, and I'd like to not mix the conversations* otherwise it will be confusing. 

Thanks

* at least not until we can verify the issue has the same  root cause :)

**Comment by christosts** [comment]
> @christosts iirc traces were taken without surface workaround

Quick check: the anr traces show the players are blocked in this [line](https://github.com/google/ExoPlayer/blob/r2.15.1/library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java#L546) (link taken from 2.15.1), which is where the DummySurface instance is created. Can you check in your device if the line is reached or not?

**Comment by denis-bezrukov** [comment]
@christosts I can check it tomorrow when I'll be in our lab, but most likely yes, because I just checked one more trace from that day (but from the main app) and it doesn't block on this line, but on `android.media.MediaCodec.native_flush`. I sent that trace to the same email thread in case if it helps

**Comment by christosts** [comment]
In high-level, issues like this tend to be (a combination of) device-specific, codec-specific and use case-specific. In both traces we received, the player is blocked on API calls to the platform. In the second trace, as you mentioned,  a few player instances are blocked on calling MediaCodec.flush(), which is a platform API. It's not easy for the player - if not impossible - to recover from cases like that. In worst case scenarios, when the platform becomes unresponsive, the device might even need to be rebooted for the codecs to become working again.

I tried your with your patch (thanks for making it) but it did not reproduce on a couple of devices, which admittedly were relatively new ones compared to your device running Android 6. I'll try to assist debugging this, though it might not be an easy task over a comments-thread:

- Codecs: you mentioned you are using a mix of H264/HEVC content. Have you noticed anything related to what of kind of content may cause a Player to become blocked? For example, does this happen more when a player instance is playing HEVC content? For example, does Player.release() time out when the player is playing HEVC while another player is already playing HEVC? I mention HEVC because this codec might be more scarce to share compared to H264, but the same question applies to H264. What kind of content did you use in your test code? Only H264, only HEVC, or both? A test you may be able to do to narrow if this is a codec related problem, is to remove randomness from your test app (-> no shuffling of URLs, fixed num of players) and use fixed combination of content (e.g. 9 players with H264, or 9 players with HEVC, or 5 with H264 and 4 with HEVC), all starting and releasing together. I understand it's a tedious task.

- ExoPlayer version: you mentioned you started seeing this when moving from 2.13 to 2.15.1. I assume the traces are obtained when your test app is using 2.15.1. What happens if you do the same test on 2.13? Does this happen? Do you use the same mix of H264/HEVC content on both 2.13 and 2.15.1? Or did you start using HEVC with 2.15.1 for example?

**Comment by denis-bezrukov** [comment]
@christosts thank you for the update,  

>Codecs: you mentioned you are using a mix of H264/HEVC content. Have you noticed anything related to what of kind of content may cause a Player to become blocked? .....

Yep, all these are good questions, and as you said currently sample app contains a lot of random stuff, so it's hard to give the answers. Currently, in my test setup I played a mixed H264/HEVC, so it's even hard to say which codec causes this (or maybe both). I remember sometimes I was able to repro this playing a single video at a time (but keeping reinitialization at a random time), so it probably not related to playing a specific combination of content. 
I will try to
1. Play H264 and HEVC separately
2. Remove randomness as much as possible
And will back to you with the results. Is there anything I can log that would be helpful? Maybe something around MediaCodecAdapter? Is there some codec's health check that I can perform (e.g. trying to init X media codecs after a release, where X is a max available instance count)

>ExoPlayer version: you mentioned you started seeing this when moving from 2.13 to 2.15.1. I assume the traces are obtained when your test app is using 2.15.1.

Yep, they are from 2.15.1. We started observing this after upgrading to 2.15.1 in our crash tracker (we report it as a warning). Now I just checked it in our log storage and it seems this issue started to happen since 2.13 but less often (player's release timeout was introduced in 2.13, before the app was simply ANRing)
![image](https://user-images.githubusercontent.com/7599577/149936939-f537fb28-8e65-4e1a-a8b9-981c906bde3c.png)

> Do you use the same mix of H264/HEVC content on both 2.13 and 2.15.1? Or did you start using HEVC with 2.15.1 for example?

No, we used HEVC for more than a year, so we played it on both 2.13/2.15

**Comment by christosts** [comment]
> Yep, they are from 2.15.1. We started observing this after upgrading to 2.15.1 in our crash tracker (we report it as a warning). Now I just checked it in our log storage and it seems this issue started to happen since 2.13 but less often (player's release timeout was introduced in 2.13, before the app was simply ANRing)

Are these errors comparable to what the ANRs used to be before you upgraded to 2.13? That is, are the following two quantities close or equal: `ANRs before upgrading to 2.13` vs `current ANRs + release timeouts`? It's likely that the app continues to show the same overall behavior with the difference that the app is now reporting these errors in place of what it used to be ANRs

**Comment by denis-bezrukov** [comment]
>That is, are the following two quantities close or equal: ANRs before upgrading to 2.13 vs current ANRs + release timeouts?

It's hard to say, we don't collect historical data for ANRs that can be aggregated quickly

>We need more information to resolve this issue but there hasn't been an update in 14 weekdays. 

Don't close it 🙏 My plan to return to this issue next two weeks

**Comment by adamvernier** [comment]
I also hope that you don't close this. I believe this issue is very similar or the same to what we've been seeing. We have hundreds of devices deployed for a digital sign product that often plays many videos. Our Android 6 devices are particularly prone to these "codec lockups" and need to be rebooted, sometimes many times a day. I very much appreciate the support on this and hope to soon provide a separate report as requested. thanks.

**Comment by denis-bezrukov** [comment]
@christosts I'm continue investigating this issue and now I added a MediaCodecAdapter wrapper that wraps setOutputSurface/release calls and logs it (before, after, error)
Normally, I see the following logs (`[]` contains Player's identity hash): 
```
2022-02-21 07:35:20.606 7099-7706 Codec [232088673] video/hevc pre release
2022-02-21 07:35:20.677 7099-7706 Codec [232088673] video/hevc post release
2022-02-21 07:35:20.736 7099-7706 Codec [232088673] audio/mp4a-latm pre release
2022-02-21 07:35:20.746 7099-7706 Codec [232088673] audio/mp4a-latm post release
```
But when the issue occurs, the logs are the following: 
```
2022-02-21 07:39:16.654 7099-8201 Codec [153359882] video/hevc pre release
2022-02-21 07:39:19.686 7099-8201 Codec [153359882] video/hevc post release
2022-02-21 07:39:34.696 7099-7099 onPlayerError 153359882
    com.google.android.exoplayer2.ExoPlaybackException: Unexpected runtime error
    ........
     Caused by: com.google.android.exoplayer2.ExoTimeoutException: Player release timed out.
```
As you can see, hevc codec was released correctly, but then the audio codec didn't even start the release process, and then player's main thread timeouts (I use 15s for now to ensure it's not a busy main thread)

Is there anything that happens after video but before audio gets disabled?

**Comment by denis-bezrukov** [comment]
It seems it's an audio flush: 
```
2022-02-21 07:55:57.089 10320-10899 Codec [166412023] video/hevc pre release
2022-02-21 07:56:00.120 10320-10899 Codec [166412023] video/hevc post release
2022-02-21 07:56:00.231 10320-10899 Codec [166412023] audio/mp4a-latm pre flush
2022-02-21 07:56:15.138 10320-10320 onPlayerError 166412023
```

UPD: After few more tests it doesn't seem to be related to a specific mime type / codec instance, I see the same pattern across different codecs (audio/mp4a-latm, video/avc, video/hevc). That makes me thinking it's not a hardware issue, but probably an issue in platform code. Is [this](https://android.googlesource.com/platform/frameworks/av/+/refs/heads/marshmallow-release/media/libstagefright/MediaCodec.cpp) a good source for the latest mediacodec version in the Marshmallow? I also tried to compare with [nougat ](https://android.googlesource.com/platform/frameworks/av/+/refs/heads/nougat-release/media/libstagefright/MediaCodec.cpp) but don't see anything related

I've also tried AsynchronousMediaCodecAdapter but the issue is still reproducible (most likely it's expected)

**Comment by christosts** [comment]
Thank you very much for the investigation. 

> Is there anything that happens after video but before audio gets disabled?

MediaCodecAudioRenderer.onDisabled() will call MediaCodecRenderer.onDisabled() which will flush the MediaCodec.

> It seems it's an audio flush

Interesting. We knew already it was a flush but my first guess was that the MediaCodec instance for video would block rather than the audio one. There are cases for which instead of flushing, we release the codec instead, please see [here](https://github.com/google/ExoPlayer/blob/r2.15.1/library/core/src/main/java/com/google/android/exoplayer2/mediacodec/MediaCodecRenderer.java#L877-L890). It might be worth trying forcing this behavior on your device too, e.g. you can override MediaCodecAudioRenderer.flushOrReleaseCodec() to always release the codec. Note that I have not tried this.

**Comment by denis-bezrukov** [comment]
@christosts 

> We knew already it was a flush but my first guess was that the MediaCodec instance for video would block rather than the audio one.

I updated my previous reply and it seems to be a general "codec" issue, since I see the same pattern (there is a `pre flush` but no `post flush`) for both audio and video codecs. 

>It might be worth trying forcing this behavior on your device too, e.g. you can override MediaCodecAudioRenderer.flushOrReleaseCodec() to always release the codec. Note that I have not tried this.

Just tried it (for both audio and video) and it does seem to work, on the other hand release process became too long:
```
2022-02-24 05:43:19.153 15394-17694 Codec [19308084] video/avc pre release
2022-02-24 05:43:22.193 15394-17694 Codec [19308084] video/avc post release
2022-02-24 05:43:22.303 15394-17694 Codec [19308084] audio/mp4a-latm pre release
2022-02-24 05:43:25.310 15394-17694 Codec [19308084] audio/mp4a-latm post release
```
As you can see the full release process took ~6s

I am also concerned that I will do that for position reset (seek) so it will look worse. According to our tracking system, the issue happens for ~2-3% of devices so I am not sure we're ok with that tradeoff.  So I would be interested to continue investigating the issue.

**Comment by wqafzal** [comment]
is this issue resolved, or any update? we are getting many errors on firestick devices.

**Comment by nikky12** [comment]
Is this issue resolved, we are getting these issues daily on mobile phone 
For reference logs
	at j.n.b.c.q2.z(SourceFile:2)
	at j.n.b.c.v0.invoke(Unknown Source:2)
	at j.n.b.c.r4.t$c.invoke(SourceFile:5)
	at j.n.b.c.r4.t.d(SourceFile:2)
	at j.n.b.c.r4.a.run(Unknown Source:6)
	at j.n.b.c.r4.t.flushEvents(SourceFile:10)
	at j.n.b.c.r4.t.sendEvent(SourceFile:2)
	at j.n.b.c.q2.release(SourceFile:6)
	at com.google.android.exoplayer2.SimpleExoPlayer.release(SourceFile:10)
	at player.ContentPlayer.destroy(SourceFile:6)
....
...
Caused by: com.google.android.exoplayer2.ExoTimeoutException: Player release timed out.
	at j.n.b.c.q2.z(SourceFile:1)
	... 56 more

**Comment by christosts** [comment]
I'm closing this issue as it was device-specific and offered workarounds on codec release. 

@wqafzal and @nikky12 if you are seeing timeouts when releasing the player, please file new issues separately. Such issues tend to be device-specific and discussing multiple occurrences on the same comments thread can be misleading. Also, when filing new issues, please provide the information requested on the bug template. 

Thanks
